src/data/polynomial/hasse_deriv.lean

```diff
@@ -0,0 +1,220 @@
+    by_cases hkn : k ≤ n,
+    { rw [← nat.sub_eq_iff_eq_add hkn] at hnik, rw [if_neg hnik] },
+    { push_neg at hkn, rw [nat.choose_eq_zero_of_lt hkn, nat.cast_zero, zero_mul, if_t_t] } }
+end
+
+lemma hasse_deriv_C (r : R) (hk : 0 < k) : hasse_deriv k (C r) = 0 :=
+by rw [← monomial_zero_left, hasse_deriv_monomial, nat.choose_eq_zero_of_lt hk,
+    nat.cast_zero, zero_mul, monomial_zero_right]
+
+lemma hasse_deriv_apply_one (hk : 0 < k) : hasse_deriv k (1 : polynomial R) = 0 :=
+by rw [← C_1, hasse_deriv_C k _ hk]
+
+lemma hasse_deriv_X (hk : 1 < k) : hasse_deriv k (X : polynomial R) = 0 :=
+by rw [← monomial_one_one_eq_X, hasse_deriv_monomial, nat.choose_eq_zero_of_lt hk,
+    nat.cast_zero, zero_mul, monomial_zero_right]
+
+lemma factorial_smul_hasse_deriv :
+  ⇑(k! • @hasse_deriv R _ k) = ((@derivative R _)^[k]) :=
+begin
+  induction k with k ih,
+  { rw [hasse_deriv_zero, factorial_zero, iterate_zero, one_smul, linear_map.id_coe], },
+  ext f n : 2,
+  rw [iterate_succ_apply', ← ih],
+  simp only [linear_map.smul_apply, coeff_smul, linear_map.map_smul_of_tower, coeff_derivative,
+    hasse_deriv_coeff, ← @choose_symm_add _ k],
+  simp only [nsmul_eq_mul, factorial_succ, mul_assoc, succ_eq_add_one, ← add_assoc,
+    add_right_comm n 1 k, ← cast_succ],
+  rw ← (cast_commute (n+1) (f.coeff (n + k + 1))).eq,
+  simp only [← mul_assoc], norm_cast, congr' 2,
+  apply @cast_injective ℚ,
+  have h1 : n + 1 ≤ n + k + 1 := succ_le_succ le_self_add,
+  have h2 : k + 1 ≤ n + k + 1 := succ_le_succ le_add_self,
+  have H : ∀ (n : ℕ), (n! : ℚ) ≠ 0, { exact_mod_cast factorial_ne_zero },
+  -- why can't `field_simp` help me here?
+  simp only [cast_mul, cast_choose ℚ, h1, h2, -one_div, -mul_eq_zero,
+    succ_sub_succ_eq_sub, nat.add_sub_cancel, add_sub_cancel_left] with field_simps,
+  rw [eq_div_iff_mul_eq (mul_ne_zero (H _) (H _)), eq_comm, div_mul_eq_mul_div,
+    eq_div_iff_mul_eq (mul_ne_zero (H _) (H _))],
+  norm_cast,
+  simp only [factorial_succ, succ_eq_add_one], ring,
+end
+.
+
+lemma hasse_deriv_comp (k l : ℕ) :
+  (@hasse_deriv R _ k).comp (hasse_deriv l) = (k+l).choose k • hasse_deriv (k+l) :=
+begin
+  ext i : 2,
+  simp only [linear_map.smul_apply, comp_app, linear_map.coe_comp, smul_monomial,
+    hasse_deriv_apply, mul_one, monomial_eq_zero_iff, sum_monomial_index, mul_zero,
+    nat.sub_sub, add_comm l k],
+  rw_mod_cast nsmul_eq_mul,
+  congr' 2,
+  by_cases hikl : i < k + l,
+  { rw [choose_eq_zero_of_lt hikl, mul_zero],
+    by_cases hil : i < l,
+    { rw [choose_eq_zero_of_lt hil, mul_zero] },
+    { push_neg at hil, rw [← nat.sub_lt_right_iff_lt_add hil] at hikl,
+      rw [choose_eq_zero_of_lt hikl , zero_mul], }, },
+  push_neg at hikl, apply @cast_injective ℚ,
+  have h1 : l ≤ i     := nat.le_of_add_le_right hikl,
+  have h2 : k ≤ i - l := nat.le_sub_right_of_add_le hikl,
+  have h3 : k ≤ k + l := le_self_add,
+  have H : ∀ (n : ℕ), (n! : ℚ) ≠ 0, { exact_mod_cast factorial_ne_zero },
+  -- why can't `field_simp` help me here?
```

Comment:
What are you wanting it to do here? Perhaps explain or remove?

Reply:
Given `H`, it should be able to remove all the divisions from the goal state. But it doesn't.